In [1]:
import pandas as pd

In [2]:
mobile = pd.read_csv("mobile_price_data.csv")

Briefly explore the raw dataset

In [3]:
mobile.head()

mobile_name mobile_price mobile_color  dual_sim  \
0    Realme Narzo 10A (So White, 32 GB)       ₹8,999     So White  Dual Sim   
1     Realme Narzo 10A (So Blue, 32 GB)       ₹8,999      So Blue  Dual Sim   
2     Realme Narzo 10A (So Blue, 64 GB)       ₹9,999      So Blue  Dual Sim   
3    Realme Narzo 10A (So White, 64 GB)       ₹9,999     So White  Dual Sim   
4  Realme Narzo 10 (That Green, 128 GB)      ₹11,999   That Green  Dual Sim   

              disp_size         resolution          os  num_cores mp_speed  \
0  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
1  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
2  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
3  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
4   16.51 cm (6.5 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   

  int_memory   ram                   p_cam              f_cam  \
0      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
1      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
2      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
3      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
4     128 GB  4 GB  48MP + 8MP + 2MP + 2MP  16MP Front Camera   

                network bluetooth battery_power mob_width mob_height  \
0  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
1  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
2  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
3  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
4  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh   75.4 mm   164.4 mm   

  mob_depth mob_weight  
0   8.95 mm      195 g  
1   8.95 mm      195 g  
2   8.95 mm      195 g  
3   8.95 mm      195 g  
4      9 mm      199 g

First of all, we remove those units in the columns, by using regex

In [4]:
import re
mobile["disp_size"] = mobile["disp_size"].apply(lambda x: re.sub(r'(.+)(cm)(.+)',r'\1', x))
mobile["mp_speed"] = mobile["mp_speed"].apply(lambda x: re.sub(r'(.+)( GHz)',r'\1', x))
mobile["int_memory"] = mobile["int_memory"].apply(lambda x: re.sub(r'(.+)( GB)',r'\1', x))
mobile["ram"] = mobile["ram"].apply(lambda x: re.sub(r'(.+)( GB)',r'\1', x))
mobile["battery_power"] = mobile["battery_power"].apply(lambda x: re.sub(r'(.+)( mAh)',r'\1', x))
mobile["mob_width"] = mobile["mob_width"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_height"] = mobile["mob_height"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_depth"] = mobile["mob_depth"].apply(lambda x: re.sub(r'(.+)( mm)',r'\1', x))
mobile["mob_weight"] = mobile["mob_weight"].apply(lambda x: re.sub(r'(.+)( g)',r'\1', x)) 

Splitting "mobile_name" into two columns: "mobile_brand" and "model_name"

In [5]:
mobile[['mobile_brand', 'model_name']] = mobile['mobile_name'].str.split(' ', 1, expand=True)
mobile.drop('mobile_name', inplace=True, axis=1)

mobile["model_name"] = mobile["model_name"].apply(lambda x: re.sub(r'(.*?) \((.*)',r'\1', x))

Splitting "resolution", "f_cam" and "p_cam", for the ease of data analysis

In [6]:
mobile[['resolution_width', 'resolution_height']] = mobile['resolution'].str.split(r'x|×|\*', 1, expand=True)
mobile.drop('resolution', inplace=True, axis=1)
mobile["resolution_height"] = mobile["resolution_height"].astype(str)
mobile["resolution_height"] = mobile["resolution_height"].apply(lambda x: re.sub(r'(\d+)(.*)',r'\1', x))

mobile[['f_cam_1', 'f_cam_2']] = mobile['f_cam'].str.split('+', 1, expand=True)
mobile.drop('f_cam', inplace=True, axis=1)
mobile["f_cam_1"] = mobile["f_cam_1"].astype(str)
mobile["f_cam_2"] = mobile["f_cam_2"].astype(str)
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["f_cam_1"] = mobile["f_cam_1"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["f_cam_2"] = mobile["f_cam_2"].apply(lambda x: re.sub(r'(\D+)',r'', x))

mobile[['p_cam_1', 'p_cam_2', 'p_cam_3', 'p_cam_4']] = mobile['p_cam'].str.split('+', 4, expand=True)
mobile.drop('p_cam', inplace=True, axis=1)
mobile["p_cam_1"] = mobile["p_cam_1"].astype(str)
mobile["p_cam_2"] = mobile["p_cam_2"].astype(str)
mobile["p_cam_3"] = mobile["p_cam_3"].astype(str)
mobile["p_cam_4"] = mobile["p_cam_4"].astype(str)
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_1"] = mobile["p_cam_1"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_2"] = mobile["p_cam_2"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_3"] = mobile["p_cam_3"].apply(lambda x: re.sub(r'(\D+)',r'', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\D+)(\d+)(\D+)',r'\2', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\d+)(\D+)',r'\1', x))
mobile["p_cam_4"] = mobile["p_cam_4"].apply(lambda x: re.sub(r'(\D+)',r'', x))

For the "network" column, we just keep the newer and better network type, i.e.: **5G>4G VOLTE>4G>3G>2G**

In [7]:
list1 = [] 
for x in mobile["network"]:
    if "5G" in x:
        list1.append("5G")
    elif "4G VOLTE" in x:
        list1.append("4G VOLTE")
    elif "4G" in x:
        list1.append("4G")
    elif "3G" in x:
        list1.append("3G")
    elif "2G" in x:
        list1.append("2G")
mobile["network"] = list1
mobile.rename(columns={'network': 'network_top'}, inplace=True)

Briefly explore our cleaned dataset, to check whether there is any error or mistake during our cleaning process

In [8]:
mobile.head()

mobile_price mobile_color  dual_sim disp_size          os  num_cores  \
0       ₹8,999     So White  Dual Sim    16.56   Android 10  Octa Core   
1       ₹8,999      So Blue  Dual Sim    16.56   Android 10  Octa Core   
2       ₹9,999      So Blue  Dual Sim    16.56   Android 10  Octa Core   
3       ₹9,999     So White  Dual Sim    16.56   Android 10  Octa Core   
4      ₹11,999   That Green  Dual Sim    16.51   Android 10  Octa Core   

  mp_speed int_memory ram network_top  ... mobile_brand model_name  \
0        2         32   3    4G VOLTE  ...       Realme  Narzo 10A   
1        2         32   3    4G VOLTE  ...       Realme  Narzo 10A   
2        2         64   4    4G VOLTE  ...       Realme  Narzo 10A   
3        2         64   4    4G VOLTE  ...       Realme  Narzo 10A   
4        2        128   4    4G VOLTE  ...       Realme   Narzo 10   

  resolution_width resolution_height f_cam_1 f_cam_2 p_cam_1 p_cam_2 p_cam_3  \
0            1600                720       5              12       2       2   
1            1600                720       5              12       2       2   
2            1600                720       5              12       2       2   
3            1600                720       5              12       2       2   
4            1600                720      16              48       8       2   

  p_cam_4  
0          
1          
2          
3          
4       2  

[5 rows x 26 columns]

In [9]:
mobile.sample(10)

mobile_price   mobile_color  dual_sim disp_size                os  \
223      ₹16,990   Skyline Blue  Dual Sim    16.21      Android Pie 9   
71       ₹27,999   Arctic White  Dual Sim    16.69         Android 10   
394      ₹12,990          Black  Dual Sim    15.75   Android Oreo 8.1   
343      ₹22,999     Steel Blue  Dual Sim     15.7   Android Oreo 8.1   
17       ₹41,999       Rust Red  Dual Sim    16.36         Android 10   
109       ₹9,799  Emerald Green  Dual Sim     15.8      Android Pie 9   
177       ₹6,700           Blue  Dual Sim     12.7   Android Oreo 8.1   
248      ₹15,990  Nebula Purple  Dual Sim     15.8   Android Oreo 8.1   
247       ₹8,900    Matte Black  Dual Sim    15.24   Android Oreo 8.1   
110       ₹9,799       Ruby Red  Dual Sim     15.8      Android Pie 9   

     num_cores mp_speed int_memory ram network_top  ... mobile_brand  \
223  Octa Core        2        128   4          4G  ...         Vivo   
71   Octa Core     2.96        128   8    4G VOLTE  ...       Realme   
394  Octa Core      2.3         32   2    4G VOLTE  ...         OPPO   
343  Octa Core      2.8        256   8    4G VOLTE  ...         POCO   
17   Octa Core      2.8        128   8          5G  ...       Realme   
109  Octa Core     1.95         64   4    4G VOLTE  ...        Redmi   
177  Quad Core      1.4         16   1    4G VOLTE  ...        Redmi   
248  Octa Core        2         32   4    4G VOLTE  ...         Vivo   
247  Quad Core      1.4         16   2    4G VOLTE  ...         Vivo   
110  Octa Core     1.95         64   4    4G VOLTE  ...        Redmi   

       model_name resolution_width resolution_height f_cam_1 f_cam_2 p_cam_1  \
223            S1            1080               2340      32              16   
71   X3 SuperZoom            2400               1080      32       8      64   
394           A5s            1520                720       8              13   
343            F1            2160               1080      20              12   
17        X50 Pro            1080               2400      32       8      64   
109             8            1520                720       8              12   
177            Go            1280                720       5               8   
248           Y93            1520                720       8              13   
247          Y71i            1440                720       5               8   
110             8            1520                720       8              12   

    p_cam_2 p_cam_3 p_cam_4  
223       8       2          
71        8       8       2  
394       2                  
343       5                  
17       12       8       2  
109       2                  
177                          
248       2                  
247                          
110       2                  

[10 rows x 26 columns]

Remove some extreme data, and fixing some minor errors that was made during the cleaning process

In [10]:
mobile.loc[230,['p_cam_1']] = "48"
mobile.loc[230,['p_cam_2']] = "8"
mobile.loc[230,['p_cam_3']] = "5"
mobile.loc[234,['p_cam_1']] = "48"
mobile.loc[234,['p_cam_2']] = "8"
mobile.loc[234,['p_cam_3']] = "5"


mobile = mobile.loc[mobile["mobile_brand"] != "MI3"]

We convert the "mobile_price", from ***Indian Rupee*** into ***Singapore Dollar***. Using the currency rate on 11 Mar 2022.

In [11]:
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: re.sub(r'(.*)\,(.*)',r'\1\2', x))
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: re.sub(r'.(\d+)',r'\1', x))
mobile["mobile_price"] = pd.to_numeric(mobile["mobile_price"])
mobile["mobile_price"] = mobile["mobile_price"].apply(lambda x: round(x*0.0178), x)

In [12]:
mobile.sample(10)

mobile_price          mobile_color  dual_sim disp_size  \
6             267         Crystal Green  Dual Sim     16.0    
378           178                 Black  Dual Sim     15.8    
91            534          Carbon Black  Dual Sim    16.23    
433           338                  Grey  Dual Sim    13.97    
423           302        Waterfall Grey  Dual Sim    16.51    
311           205          Forest Green  Dual Sim    16.59    
171           249                 Black  Dual Sim    15.21    
231           525  Midnight Ocean Black  Dual Sim    16.36    
457           249          Marble Green  Dual Sim    16.59    
87            267         Not just Blue  Dual Sim    15.44    

                       os  num_cores mp_speed int_memory ram network_top  ...  \
6         Android Pie 9.0  Octa Core      2.3         64   6    4G VOLTE  ...   
378       Android Pie 9.0  Octa Core      1.8         32   3    4G VOLTE  ...   
91        Android Pie 9.0  Octa Core     2.84        256   8    4G VOLTE  ...   
433  Android Lollipop 5.1  Octa Core      1.4         64   4          4G  ...   
423         Android Pie 9  Octa Core      2.1        128   6    4G VOLTE  ...   
311            Android 10  Octa Core      2.3         64   4    4G VOLTE  ...   
171  Android Nougat 7.1.2  Octa Core      1.8         64   6    4G VOLTE  ...   
231       Android Pie 9.0  Octa Core        2        128   8          4G  ...   
457       Android Pie 9.0  Octa Core      2.1        128   4    4G VOLTE  ...   
87          Android Pie 9  Octa Core        2        128   6    4G VOLTE  ...   

    mobile_brand  model_name resolution_width resolution_height f_cam_1  \
6         Realme       5 Pro            2340               1080      16   
378         OPPO         A12            1520                720       5   
91         Redmi     K20 Pro            2340               1080      20   
433         OPPO         F1S            1280                720      16   
423         OPPO     F11 Pro            2340               1080      16   
311      Infinix      S5 Pro            2340               1080      16   
171        Redmi  Note 5 Pro            2160               1080      20   
231         Vivo      V17Pro            2400               1080      32   
457         OPPO         F11            1080               2340      16   
87            Mi          A3            1560                720      32   

    f_cam_2 p_cam_1 p_cam_2 p_cam_3 p_cam_4  
6                48       8       2       2  
378              13       2                  
91               48      13       8          
433              13                          
423              48       5                  
311              48       2                  
171              12       5                  
231       8      48      13       8       2  
457              48       5                  
87               48       8       2          

[10 rows x 26 columns]

Remove unnecessary data in the "os" column (i.e. the .0 suffix), and renamed ***Android Pie 10*** to ***Android 10***

In [13]:
mobile.replace("Android Pie 9.0", "Android Pie 9", inplace=True)
mobile.replace("Android Pie 10", "Android 10", inplace=True) #Android 10 was initially named as Android Q, but the name was dropped after release
mobile.replace("Android Oreo 8.1.0", "Android Oreo 8.1", inplace=True)
mobile.replace("Android Oreo 8.0", "Android Oreo 8", inplace=True)

Last but not least, we simplify the "mobile_color" column, just keep the necessary information

In [14]:
mobile["mobile_color"] = mobile["mobile_color"].apply(lambda x: re.sub(r'(.*) ',r'', x))
mobile["mobile_color"].replace(["Edition", "Mist", "Sea", "Wave", "Brick", "Greener"], ["Black", "Radiant Mist", "Black", "Blue", "Red", "Green"] , inplace=True)

Finally, we rearrange our columns, to make it clearer during the interpretation process

In [15]:
list(mobile.columns.values)

['mobile_price',
 'mobile_color',
 'dual_sim',
 'disp_size',
 'os',
 'num_cores',
 'mp_speed',
 'int_memory',
 'ram',
 'network_top',
 'bluetooth',
 'battery_power',
 'mob_width',
 'mob_height',
 'mob_depth',
 'mob_weight',
 'mobile_brand',
 'model_name',
 'resolution_width',
 'resolution_height',
 'f_cam_1',
 'f_cam_2',
 'p_cam_1',
 'p_cam_2',
 'p_cam_3',
 'p_cam_4']

In [16]:
mobile = mobile[[
 'mobile_brand',
 'model_name',
 'mobile_color',
 'dual_sim',
 'disp_size',
 'os',
 'num_cores',
 'mp_speed',
 'int_memory',
 'ram',
 'network_top',
 'bluetooth',
 'battery_power',
 'mob_width',
 'mob_height',
 'mob_depth',
 'mob_weight',
 'resolution_width',
 'resolution_height',
 'f_cam_1',
 'f_cam_2',
 'p_cam_1',
 'p_cam_2',
 'p_cam_3',
 'p_cam_4',
 'mobile_price']]

In [17]:
mobile.shape

(458, 26)

Now the dataset consists of ***458*** rows and ***26*** columns. We export the cleaned dataset into a csv file to be used during the EDA process

In [18]:
mobile.to_csv("mobile_data_new.csv",mode='w+' )